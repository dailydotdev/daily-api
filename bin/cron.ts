import '../src/config';
import cron from '../src/cron';
import { createOrGetConnection } from '../src/db';

const key = process.argv[2];
const selected = cron.get(key);
if (!selected) {
  console.error(`cron "${key}" does not exist!`);
  process.exit(-1);
}

console.log(`starting ${key}`);
createOrGetConnection()
  .then((con) => selected.handler(con, ...process.argv.slice(3)))
  .then(() => {
    console.log('done');
    process.exit();
  })
  .catch((err) => {
    console.error(err);
    process.exit(-1);
  });
