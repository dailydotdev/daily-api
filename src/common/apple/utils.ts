import {
  NotificationTypeV2,
  Subtype,
  type Environment,
  type JWSRenewalInfoDecodedPayload,
  type JWSTransactionDecodedPayload,
  type ResponseBodyV2DecodedPayload,
  type SignedDataVerifier,
} from '@apple/app-store-server-library';
import { logger } from '../../logger';
import { SubscriptionProvider, type User } from '../../entity/user/User';
import { isNullOrUndefined } from '../object';
import {
  AnalyticsEventName,
  sendAnalyticsEvent,
} from '../../integrations/analytics';
import {
  AppleTransactionType,
  certificatesToLoad,
  productIdToCycle,
} from './types';
import { readFile } from 'fs/promises';
import { isTest } from '../utils';

export const verifyAndDecodeAppleSignedData = async ({
  notification,
  verifier,
}: {
  notification: ResponseBodyV2DecodedPayload;
  environment: Environment;
  verifier: SignedDataVerifier;
}): Promise<{
  transactionInfo?: JWSTransactionDecodedPayload;
  renewalInfo?: JWSRenewalInfoDecodedPayload;
}> => {
  let renewalInfo: JWSRenewalInfoDecodedPayload | undefined = undefined;
  let transactionInfo: JWSTransactionDecodedPayload | undefined = undefined;

  if (!isNullOrUndefined(notification.data?.signedRenewalInfo)) {
    renewalInfo = await verifier.verifyAndDecodeRenewalInfo(
      notification.data.signedRenewalInfo,
    );
  }

  if (!isNullOrUndefined(notification.data?.signedTransactionInfo)) {
    transactionInfo = await verifier.verifyAndDecodeTransaction(
      notification.data.signedTransactionInfo,
    );
  }

  return {
    transactionInfo,
    renewalInfo,
  };
};

export const logAppleAnalyticsEvent = async (
  transactionInfo: JWSTransactionDecodedPayload,
  renewalInfo: JWSRenewalInfoDecodedPayload,
  eventName: AnalyticsEventName,
  user: User,
  currencyInUSD: number,
) => {
  if (!transactionInfo || isTest) {
    return;
  }

  const cycle =
    productIdToCycle[
      renewalInfo?.autoRenewProductId as keyof typeof productIdToCycle
    ];
  const cost = renewalInfo?.renewalPrice;

  const extra = {
    payment: SubscriptionProvider.AppleStoreKit,
    cycle,
    cost: currencyInUSD,
    currency: 'USD',
    localCost: cost ? cost / 1000 : undefined,
    localCurrency: transactionInfo?.currency,
    payout: {
      total: currencyInUSD * 100,
      grandTotal: currencyInUSD * 100,
      currencyCode: 'USD',
    },
  };

  await sendAnalyticsEvent([
    {
      event_name: eventName,
      event_timestamp: new Date(transactionInfo?.signedDate || ''),
      event_id: transactionInfo.appTransactionId,
      app_platform: 'api',
      user_id: user.id,
      extra: JSON.stringify(extra),
    },
  ]);
};

export const getAnalyticsEventFromAppleNotification = (
  notificationType: ResponseBodyV2DecodedPayload['notificationType'],
  subtype?: ResponseBodyV2DecodedPayload['subtype'],
): AnalyticsEventName | null => {
  switch (notificationType) {
    case NotificationTypeV2.SUBSCRIBED:
    case NotificationTypeV2.DID_RENEW:
    case NotificationTypeV2.ONE_TIME_CHARGE:
      return AnalyticsEventName.ReceivePayment;
    case NotificationTypeV2.DID_CHANGE_RENEWAL_PREF:
      return AnalyticsEventName.ChangeBillingCycle;
    case NotificationTypeV2.DID_CHANGE_RENEWAL_STATUS: // Disable/Enable Auto-Renew
      return subtype === Subtype.AUTO_RENEW_ENABLED
        ? null
        : AnalyticsEventName.CancelSubscription;
    case NotificationTypeV2.DID_FAIL_TO_RENEW:
      // When user fails to renew and there is no grace period
      if (isNullOrUndefined(subtype)) {
        return AnalyticsEventName.CancelSubscription;
      }
    default:
      return null;
  }
};

export const loadAppleRootCAs = async (): Promise<Buffer[]> => {
  const appleRootCAs: Buffer[] = await Promise.all(
    certificatesToLoad.map(async (certPath) => await readFile(certPath)),
  );

  logger.debug(`Loaded ${appleRootCAs.length} Apple's Root CAs`);
  return appleRootCAs;
};

export const getAppleTransactionType = ({
  transactionInfo,
}: {
  transactionInfo: JWSTransactionDecodedPayload;
}): AppleTransactionType | null => {
  switch (transactionInfo.type) {
    case 'Auto-Renewable Subscription':
      return AppleTransactionType.AutoRenewableSubscription;
    case 'Non-Consumable':
      return AppleTransactionType.NonConsumable;
    case 'Consumable':
      return AppleTransactionType.Consumable;
    case 'Non-Renewing Subscription':
      return AppleTransactionType.NonRenewingSubscription;
    default:
      return null;
  }
};
