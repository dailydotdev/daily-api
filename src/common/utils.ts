import { startOfISOWeek, endOfISOWeek } from 'date-fns';
import { zonedTimeToUtc } from 'date-fns-tz';
import { snakeCase } from 'lodash';

const REMOVE_SPECIAL_CHARACTERS_REGEX = /[^a-zA-Z0-9-_#.]/g;

export const ghostUser = {
  id: '404',
  username: 'ghost',
  name: 'Deleted user',
};

interface GetTimezonedIsoWeekProps {
  date: Date;
  timezone: string;
}
export const getTimezonedStartOfISOWeek = ({
  date,
  timezone,
}: GetTimezonedIsoWeekProps): Date => {
  return zonedTimeToUtc(startOfISOWeek(date), timezone);
};

export const getTimezonedEndOfISOWeek = ({
  date,
  timezone,
}: GetTimezonedIsoWeekProps): Date => {
  return zonedTimeToUtc(endOfISOWeek(date), timezone);
};

export const updateFlagsStatement = <Entity extends { flags: object }>(
  update: Partial<Entity['flags']>,
): (() => string) => {
  return () => `flags || '${JSON.stringify(update)}'`;
};

/**
 * Remove special characters from a string
 * Only allows alphanumeric, -, _, # and . characters
 *
 * @param text
 */
export const removeSpecialCharacters = (text: string): string => {
  return text.replace(REMOVE_SPECIAL_CHARACTERS_REGEX, '');
};

/**
 * Remove duplicate values from an array
 *
 * @param array
 */
export const uniqueifyArray = <T>(array: T[]): T[] => {
  return [...new Set(array)];
};

/**
 * Remove empty values from an array
 *
 * @param array
 */
export const removeEmptyValues = <T>(array: T[]): T[] => array.filter(Boolean);

/**
 * Check if the current environment is production
 *
 * @returns boolean
 */
export const isProd = process.env.NODE_ENV === 'production';

export const parseDate = (date: string | Date): Date | undefined => {
  if (!date) {
    return undefined;
  }

  const parsedDate = new Date(date);

  if (Number.isNaN(parsedDate.getTime())) {
    return undefined;
  }

  if (parsedDate.getTime() < 0) {
    return undefined;
  }

  return parsedDate;
};

export const toGQLEnum = (value: Record<string, string>, name: string) => {
  return `enum ${name} { ${Object.values(value).join(' ')} }`;
};

export function camelCaseToSnakeCase(
  obj: Record<string, unknown>,
): Record<string, unknown> {
  const snakeObj: Record<string, unknown> = {};
  for (const key in obj) {
    snakeObj[snakeCase(key)] = obj[key];
  }
  return snakeObj;
}

export function debeziumTimeToDate(time: number): Date {
  return new Date(Math.floor(time / 1000));
}
