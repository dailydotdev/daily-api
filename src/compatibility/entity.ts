import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
