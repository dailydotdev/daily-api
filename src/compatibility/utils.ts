import { GraphQLError } from 'graphql';
import { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';

export interface GraphqlPayload {
  query: string;
  operationName?: string;
  variables?: Record<string, unknown>;
}

export const injectGraphql = async (
  fastify: FastifyInstance,
  payload: GraphqlPayload,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extractResponse: (obj: Record<string, unknown>) => any,
  req: FastifyRequest,
  res: FastifyReply,
): Promise<FastifyReply> => {
  const reqHeaders = {
    ...req.headers,
  };
  delete reqHeaders['content-length'];
  const graphqlRes = await fastify.inject({
    method: 'POST',
    url: '/graphql',
    headers: reqHeaders,
    payload,
  });

  if (graphqlRes.statusCode !== 200) {
    req.log.warn(
      { graphqlResponse: graphqlRes.body },
      'unexpected status code when injecting graphql request',
    );
    return res.status(graphqlRes.statusCode).send(graphqlRes.rawPayload);
  }

  const json = await graphqlRes.json();
  const errors = json['errors'] as GraphQLError[];
  const code = errors?.[0]?.extensions?.code;
  if (code === 'UNAUTHENTICATED') {
    return res.status(401).send();
  }
  if (code === 'FORBIDDEN') {
    return res.status(403).send();
  }
  if (code === 'VALIDATION_ERROR' || code === 'GRAPHQL_VALIDATION_FAILED') {
    return res.status(400).send();
  }
  if (code === 'NOT_FOUND') {
    return res.status(404).send();
  }
  if (code || errors?.length) {
    req.log.warn(
      { graphqlResponse: json },
      'unexpected graphql error when injecting graphql request',
    );
    return res.status(500).send();
  }

  const resBody = extractResponse(json);
  return res.status(resBody ? 200 : 204).send(resBody);
};
