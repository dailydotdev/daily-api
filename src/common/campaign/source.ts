import {
  CampaignSource,
  CampaignState,
  CampaignType,
  Post,
  Source,
  type ConnectionManager,
} from '../../entity';

import type { AuthContext } from '../../Context';
import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import {
  ensureSourcePermissions,
  SourcePermissions,
} from '../../schema/sources';
import { ValidationError } from 'apollo-server-errors';
import { randomUUID } from 'crypto';
import { addDays } from 'date-fns';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { usdToCores } from '../number';
import { systemUser, updateFlagsStatement } from '../utils';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationBuilder } from '../../notifications/builder';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import type { NotificationCampaignSourceContext } from '../../notifications';

export const validateSquadBoostPermissions = async (
  ctx: AuthContext,
  sourceId: string,
) => {
  const source = await ensureSourcePermissions(
    ctx,
    sourceId,
    SourcePermissions.BoostSquad,
  );

  if (source.flags.campaignId) {
    throw new ValidationError('Source already has a campaign');
  }

  return source;
};

export const getSourceTags = async (
  con: ConnectionManager,
  sourceId: string,
): Promise<string[]> => {
  const result = await con.getRepository(Post).query<{ tag: string }[]>(
    `
      WITH recent_posts AS (
        SELECT id, "sharedPostId"
        FROM post
        WHERE "sourceId" = $1
        ORDER BY "createdAt" DESC
        LIMIT 30
      )
      SELECT DISTINCT ps.keyword AS tag
      FROM post_keyword ps
      INNER JOIN recent_posts rp
        ON ps."postId" = rp.id
        OR ps."postId" = rp."sharedPostId"
      WHERE ps.status = 'allow'
      LIMIT 30;
    `,
    [sourceId],
  );

  return result.map(({ tag }) => tag.trim()).filter(Boolean);
};

export const startCampaignSource = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value } = args;
  const source = await validateSquadBoostPermissions(ctx, value);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignSource).save(
      manager.getRepository(CampaignSource).create({
        id,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: source.id,
        state: CampaignState.Active,
        endedAt,
        sourceId: source.id,
        type: CampaignType.Squad,
      }),
    );

    const campaignId = campaign.id;
    const last30tags = await getSourceTags(manager, source.id);
    const finalTags = last30tags.length > 3 ? last30tags : []; // when it is 3 or below, we set global targeting

    await skadiApiClientV2.startCampaign(campaign, finalTags);

    await manager
      .getRepository(Source)
      .update(
        { id: source.id },
        { flags: updateFlagsStatement<Source>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Squad Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignSource = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(CampaignSource)
      .update({ id: campaignId }, { state: CampaignState.Cancelled });

    await manager
      .getRepository(Source)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Source>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Squad Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.SquadBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const generateCampaignSquadNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignSourceContext,
) => {
  const { campaign, source, event, user } = ctx;

  return builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`)
    .avatarSource(source);
};

export const generateCampaignSquadEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(CampaignSource)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  const source = await con
    .getRepository(Source)
    .findOneByOrFail({ id: campaign.sourceId });

  return {
    start_date: formatMailDate(campaign.createdAt),
    end_date: formatMailDate(campaign.endedAt),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    source_image: source.image,
    source_handle: source.handle,
    source_name: source.name,
  };
};
