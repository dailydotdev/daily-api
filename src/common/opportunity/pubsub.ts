import { DataSource, type EntityManager } from 'typeorm';
import { FastifyBaseLogger } from 'fastify';
import {
  CandidateAcceptedOpportunityMessage,
  CandidatePreferenceUpdated,
  MatchedCandidate,
  OpportunityMessage,
  RecruiterAcceptedCandidateMatchMessage,
  Salary,
  UserCV,
} from '@dailydotdev/schema';
import { demoCompany, triggerTypedEvent } from '../../common';
import { getSecondsTimestamp } from '../date';
import { UserCandidatePreference } from '../../entity/user/UserCandidatePreference';
import { ChangeObject } from '../../types';
import { OpportunityMatch } from '../../entity/OpportunityMatch';
import { OpportunityJob } from '../../entity/opportunities/OpportunityJob';
import { UserCandidateKeyword } from '../../entity/user/UserCandidateKeyword';
import { queryReadReplica } from '../queryReadReplica';
import { ContentPreference } from '../../entity/contentPreference/ContentPreference';
import {
  ContentPreferenceStatus,
  ContentPreferenceType,
} from '../../entity/contentPreference/types';
import { ContentPreferenceOrganization } from '../../entity/contentPreference/ContentPreferenceOrganization';

const fetchCandidateKeywords = async (
  manager: EntityManager,
  candidatePreference: UserCandidatePreference | null,
): Promise<Array<string>> => {
  if (!candidatePreference) {
    return [];
  }

  // Fetch custom keywords if enabled
  if (candidatePreference.customKeywords) {
    const customKeywords = await manager
      .getRepository(UserCandidateKeyword)
      .findBy({
        userId: candidatePreference.userId,
      });

    return customKeywords.map((k) => k.keyword);
  }

  // Otherwise fetch keywords from content preferences
  const feedKeywords = await manager
    .createQueryBuilder()
    .select(`"keywordId"`, 'keyword')
    .from(ContentPreference, 'cpk')
    .where(`cpk."feedId" = :userId`, {
      userId: candidatePreference.userId,
    })
    .andWhere('cpk.type = :contentPreferenceType', {
      contentPreferenceType: ContentPreferenceType.Keyword,
    })
    .andWhere('cpk.status != :contentPreferenceStatus', {
      contentPreferenceStatus: ContentPreferenceStatus.Blocked,
    })
    .getRawMany<{ keyword: string }>();

  return feedKeywords.map((k) => k.keyword);
};

export const notifyOpportunityMatchAccepted = async ({
  con,
  logger,
  data,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  data: ChangeObject<OpportunityMatch> | null;
}) => {
  if (!data) {
    logger.warn('No data provided for opportunity match accepted notification');
    return;
  }

  const { match, candidatePreference, keywords } = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const [match, candidatePreference] = await Promise.all([
        queryRunner.manager.getRepository(OpportunityMatch).findOneBy({
          opportunityId: data.opportunityId,
          userId: data.userId,
        }),
        queryRunner.manager
          .getRepository(UserCandidatePreference)
          .findOneBy({ userId: data.userId }),
      ]);

      const keywords = await fetchCandidateKeywords(
        queryRunner.manager,
        candidatePreference,
      );

      return { match, candidatePreference, keywords };
    },
  );

  if (!match) {
    logger.warn(
      { opportunityId: data.opportunityId, userId: data.userId },
      'Opportunity match not found for accepted notification',
    );
    return;
  }

  if (!candidatePreference) {
    logger.warn(
      { userId: data.userId },
      'Candidate preference not found for user accepting opportunity',
    );
    return;
  }

  const message = new CandidateAcceptedOpportunityMessage({
    opportunityId: match.opportunityId,
    userId: match.userId,
    createdAt: getSecondsTimestamp(match.createdAt),
    updatedAt: getSecondsTimestamp(match.updatedAt),
    screening: match.screening,
    candidatePreference: {
      ...candidatePreference,
      salaryExpectation: new Salary({
        min: candidatePreference.salaryExpectation?.min
          ? BigInt(candidatePreference.salaryExpectation.min)
          : undefined,
        period: candidatePreference.salaryExpectation?.period ?? undefined,
      }),
      cv: new UserCV({
        ...candidatePreference.cv,
        lastModified:
          getSecondsTimestamp(candidatePreference.cv.lastModified || 0) ||
          undefined,
      }),
      updatedAt: getSecondsTimestamp(candidatePreference.updatedAt),
      keywords: keywords,
    },
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.candidate-accepted-opportunity',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send opportunity match accepted event',
    );
  }
};

export const notifyRecruiterCandidateMatchAccepted = async ({
  con,
  logger,
  data,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  data: ChangeObject<OpportunityMatch> | null;
}) => {
  if (!data) {
    logger.warn(
      'No data provided for candidate opportunity match accepted notification',
    );
    return;
  }

  const match = await queryReadReplica(con, async ({ queryRunner }) => {
    return queryRunner.manager.getRepository(OpportunityMatch).findOne({
      select: ['opportunityId', 'userId'],
      where: { opportunityId: data.opportunityId, userId: data.userId },
    });
  });

  if (!match) {
    logger.warn(
      { opportunityId: data.opportunityId, userId: data.userId },
      'Opportunity match not found for recruiter accepted candidate notification',
    );
    return;
  }

  const message = new RecruiterAcceptedCandidateMatchMessage({
    opportunityId: match.opportunityId,
    userId: match.userId,
    createdAt: getSecondsTimestamp(match.createdAt),
    updatedAt: getSecondsTimestamp(match.updatedAt),
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.recruiter-accepted-candidate-match',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send recruiter accepted candidate match event',
    );
  }
};

export const notifyJobOpportunity = async ({
  con,
  logger,
  isUpdate,
  opportunityId,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  isUpdate: boolean;
  opportunityId: string;
}) => {
  const topicName = isUpdate
    ? 'api.v1.opportunity-updated'
    : 'api.v1.opportunity-added';

  const [opportunity, organization, keywords] = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const opportunity = await queryRunner.manager
        .getRepository(OpportunityJob)
        .findOneOrFail({
          where: { id: opportunityId },
          relations: {
            organization: true,
            keywords: true,
          },
        });

      const [organization, keywords] = await Promise.all([
        opportunity.organization,
        opportunity.keywords,
      ]);

      return [opportunity, organization, keywords];
    },
  );

  if (!organization) {
    logger.warn(
      {
        opportunityId: opportunity.id,
        organizationId: opportunity.organizationId,
      },
      'opportunity has no organization, skipping',
    );
    return;
  }

  /**
   * Demo logic: if the company is the demo company we can omit using Gondul and simply return the users from that company as matched candidates
   */
  if (organization.id === demoCompany.id) {
    const members = await con
      .getRepository(ContentPreferenceOrganization)
      .find({
        select: ['userId'],
        where: { organizationId: organization.id },
      });
    for (const { userId } of members) {
      await triggerTypedEvent(
        logger,
        'gondul.v1.candidate-opportunity-match',
        new MatchedCandidate({
          opportunityId,
          userId,
          matchScore: 0.87,
          reasoning:
            "We have noticed that you've been digging into React performance optimization and exploring payment systems lately. Your skills in TypeScript and Node.js line up directly with the core technologies this team uses. You also follow several Atlassian engineers and have shown consistent interest in project management software, which makes this role a natural fit for your trajectory.",
          reasoningShort:
            'Your skills in TypeScript and Node.js line up directly with the core technologies this team uses.',
        }),
      );
    }
    return;
  }

  const message = new OpportunityMessage({
    opportunity: {
      ...opportunity,
      createdAt: getSecondsTimestamp(opportunity.createdAt),
      updatedAt: getSecondsTimestamp(opportunity.updatedAt),
      keywords: keywords.map((k) => k.keyword),
    },
    organization: {
      ...organization,
      createdAt: getSecondsTimestamp(organization.createdAt),
      updatedAt: getSecondsTimestamp(organization.updatedAt),
    },
  });

  try {
    await triggerTypedEvent(logger, topicName, message);
  } catch (_err) {
    const err = _err as Error;
    logger.error({ err, message }, 'failed to send opportunity event');
  }
};

export const notifyCandidatePreferenceChange = async ({
  con,
  logger,
  userId,
}: {
  con: DataSource;
  logger: FastifyBaseLogger;
  userId: string;
}) => {
  const { candidatePreference, keywords } = await queryReadReplica(
    con,
    async ({ queryRunner }) => {
      const candidatePreference = await queryRunner.manager
        .getRepository(UserCandidatePreference)
        .findOneBy({ userId: userId });

      const keywords = await fetchCandidateKeywords(
        queryRunner.manager,
        candidatePreference,
      );

      return { candidatePreference, keywords };
    },
  );

  if (!candidatePreference) {
    logger.warn(
      { userId },
      'Candidate preference not found for user, skipping notification',
    );
    return;
  }

  const message = new CandidatePreferenceUpdated({
    payload: {
      ...candidatePreference,
      salaryExpectation: new Salary({
        min: candidatePreference.salaryExpectation?.min
          ? BigInt(candidatePreference.salaryExpectation.min)
          : undefined,
        period: candidatePreference.salaryExpectation?.period ?? undefined,
      }),
      cv: new UserCV({
        ...candidatePreference?.cv,
        lastModified:
          getSecondsTimestamp(candidatePreference?.cv?.lastModified || 0) ||
          undefined,
      }),
      updatedAt:
        getSecondsTimestamp(candidatePreference?.updatedAt) || undefined,
      keywords: keywords,
    },
  });

  try {
    await triggerTypedEvent(
      logger,
      'api.v1.candidate-preference-updated',
      message,
    );
  } catch (_err) {
    const err = _err as Error;
    logger.error(
      { err, message },
      'failed to send candidate preference updated event',
    );
  }
};
