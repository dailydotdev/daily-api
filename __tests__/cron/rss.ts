import { Connection, getConnection } from 'typeorm';
import { PubSub } from '@google-cloud/pubsub';
import { mocked } from 'ts-jest/utils';

import cron from '../../src/cron/rss';
import {
  expectSuccessfulCron,
  saveFixtures,
  setupStaticServer,
} from '../helpers';
import { Source, SourceFeed } from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { FastifyInstance } from 'fastify';
import appFunc from '../../src/background';

let con: Connection;
let app: FastifyInstance;
let staticApp: FastifyInstance;

let mockTopic: jest.Mock;

jest.mock('@google-cloud/pubsub', () => {
  const mockTopic = jest.fn();
  return {
    PubSub: jest.fn().mockImplementation(() => ({ topic: mockTopic })),
  };
});

beforeAll(async () => {
  mockTopic = mocked(new PubSub().topic);
  jest.clearAllMocks();
  con = await getConnection();
  app = await appFunc();
  staticApp = await setupStaticServer('rss.xml');
  return app.ready();
});

beforeEach(async () => {
  await saveFixtures(con, Source, sourcesFixture);
  await con.getRepository(SourceFeed).save({
    sourceId: 'a',
    feed: 'http://localhost:6789/rss.xml',
    lastFetched: new Date('Mon, 24 Aug 2020 14:07:41 +0000'),
  });
});

afterAll(async () => {
  await staticApp.close();
});

it('should dispatch message for every new article', async () => {
  const mockPublish = jest.fn().mockResolvedValue('');
  mockTopic.mockImplementation(() => ({ publishJSON: mockPublish }));
  await expectSuccessfulCron(app, cron, {
    feed: 'http://localhost:6789/rss.xml',
  });
  expect(mockPublish.mock.calls).toMatchSnapshot();
  const feed = await con
    .getRepository(SourceFeed)
    .findOne({ feed: 'http://localhost:6789/rss.xml' });
  expect(feed.lastFetched).toEqual(new Date('Mon, 24 Aug 2020 20:49:54 +0000'));
});
