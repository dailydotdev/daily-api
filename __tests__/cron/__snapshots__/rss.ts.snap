// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should dispatch message for every new article 1`] = `
Array [
  Array [
    Object {
      "id": "https://dev.to/mongodb/mongodb-weeklyupdate-1-guess-who-s-back-1kf0",
      "publicationId": "a",
      "publishedAt": 2020-08-24T20:49:54.000Z,
      "tags": Array [
        "mongodb",
        "programming",
        "database",
      ],
      "title": "MongoDB $weeklyUpdate #1: Guess Who's Back?",
      "updatedAt": 2020-08-24T20:49:54.000Z,
      "url": "https://dev.to/mongodb/mongodb-weeklyupdate-1-guess-who-s-back-1kf0",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/bistaastha/gsoc-with-chaoss-final-three-weeks-gj3",
      "publicationId": "a",
      "publishedAt": 2020-08-24T19:35:38.000Z,
      "tags": Array [],
      "title": "GSoC with CHAOSS - Final three weeks!",
      "updatedAt": 2020-08-24T19:35:38.000Z,
      "url": "https://dev.to/bistaastha/gsoc-with-chaoss-final-three-weeks-gj3",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/jamesncox/flexbox-rocks-2bf9",
      "publicationId": "a",
      "publishedAt": 2020-08-24T19:17:21.000Z,
      "tags": Array [
        "css",
        "beginners",
        "webdev",
        "tutorial",
      ],
      "title": "Flexbox Rocks!",
      "updatedAt": 2020-08-24T19:17:21.000Z,
      "url": "https://dev.to/jamesncox/flexbox-rocks-2bf9",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/freddilynne/finally-understanding-javascript-50hb",
      "publicationId": "a",
      "publishedAt": 2020-08-24T19:01:20.000Z,
      "tags": Array [
        "codenewbie",
        "javascript",
        "beginners",
        "books",
      ],
      "title": "Finally Understanding JavaScript",
      "updatedAt": 2020-08-24T19:01:20.000Z,
      "url": "https://dev.to/freddilynne/finally-understanding-javascript-50hb",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/fimrandev/things-i-wish-i-knew-when-i-started-programming-361e",
      "publicationId": "a",
      "publishedAt": 2020-08-24T18:38:36.000Z,
      "tags": Array [
        "computerscience",
        "beginners",
        "codenewbie",
        "programming",
      ],
      "title": "üïî Things I Wish I Knew When I Started Programming",
      "updatedAt": 2020-08-24T18:38:36.000Z,
      "url": "https://dev.to/fimrandev/things-i-wish-i-knew-when-i-started-programming-361e",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/valentinaperic/writing-if-statements-with-the-ternary-operator-37j",
      "publicationId": "a",
      "publishedAt": 2020-08-24T16:47:10.000Z,
      "tags": Array [
        "javascript",
        "tutorial",
        "webdev",
        "refactorit",
      ],
      "title": "Writing If Statements with the Ternary Operator",
      "updatedAt": 2020-08-24T16:47:10.000Z,
      "url": "https://dev.to/valentinaperic/writing-if-statements-with-the-ternary-operator-37j",
    },
  ],
  Array [
    Object {
      "id": "https://dev.to/sarthology/what-are-you-planning-to-make-for-hacktober-fest-2020-4bk8",
      "publicationId": "a",
      "publishedAt": 2020-08-24T15:23:08.000Z,
      "tags": Array [
        "watercooler",
        "discuss",
      ],
      "title": "What are you planning to make for \\"Hacktober Fest 2020\\"? üë©üèª‚ÄçüíªüöÄ",
      "updatedAt": 2020-08-24T15:23:08.000Z,
      "url": "https://dev.to/sarthology/what-are-you-planning-to-make-for-hacktober-fest-2020-4bk8",
    },
  ],
]
`;
